corehq/apps/es/users.py

```diff
@@ -73,9 +75,51 @@ class ElasticUser(ElasticDocumentAdapter):
     def mapping(self):
         return get_adapter_mapping(self)
 
-    @classmethod
-    def from_python(cls, doc):
-        return from_dict_with_possible_id(doc)
+    def from_python(self, user):
+        """
+        :param user: an instance of ``CouchUser`` or a user dict
+        :raises ``TypeError`` user is none of the above types
+        """
+        from corehq.apps.users.models import CouchUser
+        if isinstance(user, CouchUser):
+            user_dict = user.to_json()
+        elif isinstance(user, dict):
+            user_dict = user
+        else:
+            raise TypeError(f"Unkown type {type(user)}")
+        return self.from_dict(user_dict)
+
+    def from_dict(self, user):
```

Comment:
> what do you think of always passing in a dict that can be mutated, so the decision to copy or not can be pushed up to from_python which can make a more intelligent decision? For example, no copy is necessary when the thing being passed in here came from user.to_json().

The idea crossed my mind when I was implementing it, but I held back from doing it because we would be forced to use `deepcopy` everytime. Only that would ensure that the dict passed is safe to mutate. If we give this responsibility to `from_dict` then we could judicially copy it as per our requirements.

Reply:
Making `from_dict` private [55abccf](https://github.com/dimagi/commcare-hq/pull/32660/commits/55abccf9e5d256f1eac62dc740a540169bfcc5d0)

Reply:
> Making `from_dict` private [55abccf](https://github.com/dimagi/commcare-hq/pull/32660/commits/55abccf9e5d256f1eac62dc740a540169bfcc5d0)

+1, the private method is a nice change.

Going a bit further, it seems with the new `from_python()` flow, it wouldn't be such a stretch to just remove `_from_dict()` entirely and in-line the logic inside the `from_python()` method now (example below). This is just something that comes to mind from my (possibly naive) reviewer perspective, it's completely fine if you prefer to keep them separate.

```python
    def from_python(self, doc):
        from corehq.some.models import ModelClass
        if isinstance(doc, ModelClass):
            doc_dict = doc.to_json()
        elif isinstance(doc, dict):
            doc_dict = doc
        else:
            raise TypeError(f"Invalid type: {type(doc)} (expected {ModelClass} or dict)")

        # all code in `_from_dict()` can be added here, in-line

        return doc_dict.pop("_id"), doc_dict
```